cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -252,6 +266,39 @@ struct StructFieldFunctor {
     return Status::OK();
   }
 
+  static Result<std::shared_ptr<Array>> ApplyFieldRef(KernelContext* ctx,
+                                                      const FieldRef& field_ref,
+                                                      std::shared_ptr<Array> current) {
+    if (current->type_id() != Type::STRUCT) {
+      return Status::Invalid("Not a StructArray: ", current->ToString(),
+                             "\nMaybe a bad FieldRef? ", field_ref.ToString());
+    }
+
+    if (field_ref.IsName()) {
+      const auto& array = checked_cast<const StructArray&>(*current);
+      current = array.GetFieldByName(*field_ref.name());
```

Comment:
I think `GetFieldByName` actually has the same concern as mentioned before about not taking into account top-level nulls? (so also for this case can convert to FieldPath with FindOne, and then this can be simplified a bit further) 
(if that's the case, that probably indicates the tests needs to be expanded with a case of using a named FieldRef in the test with nulls)